File: common/src/main/java/dev/obscuria/tooltips/client/element/icon/StaticIcon.java
package dev.obscuria.tooltips.client.element.icon;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import dev.obscuria.tooltips.client.element.Transform;
import dev.obscuria.tooltips.client.renderer.TooltipContext;
import net.minecraft.client.gui.GuiGraphics;

public record StaticIcon(Transform transform) implements TooltipIcon {

    public static final Codec<StaticIcon> CODEC;

    @Override
    public Codec<StaticIcon> codec() {
        return CODEC;
    }

    @Override
    public void render(GuiGraphics graphics, TooltipContext context, int x, int y) {
        pushTransform(transform, graphics, context, x, y);
        graphics.renderItem(context.stack(), 0, 0);
        popTransform(graphics);
    }

    static {
        CODEC = RecordCodecBuilder.create(codec -> codec.group(
                Transform.CODEC.fieldOf("transform").forGetter(StaticIcon::transform)
        ).apply(codec, StaticIcon::new));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/element/icon/StaticIcon.kt
package dev.obscuria.tooltips.client.element.icon

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.world.phys.Vec3
import java.util.*

class StaticIcon(
    offset: Optional<Vec3>,
    scale: Optional<Float>
) : TransformableIcon(offset, scale) {

    override fun codec(): Codec<out ITooltipIcon> = CODEC

    companion object {

        val CODEC: Codec<StaticIcon> = RecordCodecBuilder.create { codec ->
            codec.group(
                Vec3.CODEC
                    .optionalFieldOf("offset")
                    .forGetter(StaticIcon::offset),
                Codec.FLOAT
                    .optionalFieldOf("scale")
                    .forGetter(StaticIcon::scale)
            ).apply(codec, ::StaticIcon)
        }
    }
}